src/datafusion/table_provider.rs

```diff
@@ -101,22 +210,21 @@ impl TableProvider for ZarrTableProvider {
         &self,
         filters: &[&Expr],
     ) -> datafusion::error::Result<Vec<TableProviderFilterPushDown>> {
-        // TODO handle predicates on partition columns as Exact.
         Ok(filters
             .iter()
             .map(|filter| {
                 if expr_applicable_for_cols(
                     &self
-                        .table_schema
-                        .fields
+                        .options
+                        .table_partition_cols
                         .iter()
-                        .map(|field| field.name().to_string())
+                        .map(|x| x.0.clone())
                         .collect::<Vec<_>>(),
                     filter,
                 ) {
-                    TableProviderFilterPushDown::Inexact
+                    TableProviderFilterPushDown::Exact
                 } else {
-                    TableProviderFilterPushDown::Unsupported
+                    TableProviderFilterPushDown::Inexact
```

Comment:
So this change is that everything gets pushed down, but only file schema fields are inexact?

Reply:
Yes, so if all the fields in a particular filter expression are partitions, then the filter can be applied exactly (i.e. datafusion won't bother applying that filter to what your reader/stream returns after the pushdown, if I understand correctly).
For fields that are not partitions, so fields that are in the file schema as you said, the zarr reader will try to skip reading whole chunk files if possible, to avoid the I/O, but that's it, it makes no guarantee that the result of that will satisfy the filter. So, again from my understanding, that's what Inexact means here, the filter will be pushed down to the reader, but datafusion will still apply it to what the reader returns.